test/unit/ec2_udev_rules/test_manage_volume.py

```diff
@@ -31,3 +43,165 @@ def test_validate_device_name(device_name, raises):
+    assert_that(get_all_devices()).contains("/dev/" + name)
+
+
+def test_get_imdsv2_token(mocker):
+    mock = mocker.Mock()
+    mocker.patch("requests.put", mock)
+    mock.return_value.status_code = 200
+    mock.return_value.content = {"key": "value"}
+    assert_that(get_imdsv2_token()).is_equal_to({"X-aws-ec2-metadata-token": {"key": "value"}})
+
+
+@pytest.fixture
+def volume_response():
+    return {
+        "Device": "dev-01",
+        "InstanceId": "instance-01",
+        "State": "attached",
+        "VolumeId": "vol-01",
+        "DeleteOnTermination": True,
+    }
+
+
+@pytest.fixture
+def ec2(mocker):
+    mock = mocker.MagicMock()
+    mocker.patch("boto3.client", mock)
+    return mock
+
+
+@pytest.mark.parametrize(
+    ("state", "message"),
+    [("attached", ""), ("busy", "bad state"), ("detached", "bad state"), ("available", "failed to mount")],
+)
+def test_attach_volume(mocker, volume_response, state, message, ec2, capsys):
+    mocker.patch("os.popen", mocker.mock_open(read_data="sdf"))
+    mocker.patch("subprocess.check_output", return_value="xvda")
+
+    volume_response["State"] = state
+    ec2.attach_volume.return_value = volume_response
+    ec2.describe_volumes.return_value = {"Volumes": [{"Attachments": [{"State": state}]}]}
+
+    if state != "attached":
+        with pytest.raises(SystemExit) as e:
+            attach_volume(volume_response["VolumeId"], volume_response["InstanceId"], ec2)
+        assert_that(e.type).is_equal_to(SystemExit)
+        assert_that(e.value.code).is_equal_to(1)
+
+    captured = capsys.readouterr()
+    assert_that(captured.out).contains(message)
+
+
+@pytest.mark.parametrize(
+    ("state", "message"),
+    [("attached", "bad state"), ("busy", "bad state"), ("detached", "failed to detach"), ("available", "")],
+)
+def test_detach_volume(volume_response, state, message, ec2, capsys):
+    volume_response["State"] = state
+    ec2.detach_volume.return_value = volume_response
+    ec2.describe_volumes.return_value = {"Volumes": [{"State": state}]}
+
+    if state != "available":
+        with pytest.raises(SystemExit) as e:
+            detach_volume(volume_response["VolumeId"], ec2)
+        assert_that(e.type).is_equal_to(SystemExit)
+        assert_that(e.value.code).is_equal_to(1)
+
+    captured = capsys.readouterr()
+    assert_that(captured.out).contains(message)
+
+
+@pytest.mark.parametrize(
+    "volume_output",
+    [
+        {"Volumes": [{"State": "available"}]},
+        {"Volumes": [{"State": "in-use"}]},
+        {"Volumes": [{"State": "busy"}]},
+        Exception,
+    ],
+)
+def test_is_volume_available(ec2, volume_output, capsys):
+    ec2.describe_volumes.return_value = volume_output
+    is_avail = is_volume_available(ec2, 0)
+
+    if volume_output == Exception:
+        captured = capsys.readouterr()
+        assert_that(captured.out).contains("exception")
+        assert_that(is_avail).is_false()
```

Comment:
Use the True and False value as an `expected_output` parameter in your test. Then you wont have to use condition where you check if volume_output has available or not